### PR TITLE
Make hash function a Hashtable method rather than a global function.

### PR DESCRIPTION
**Pull request by ctb**
This PR moves the _hash and _revhash calls into object-specific methods.  This paves the way to
more radical refactoring of hash functions down the road to support e.g. #1490, ref #1426.

Briefly,
- create new methods hash_dna and unhash_dna on Hashtable objects;
- refactor code in subset.cc and read_aligner.cc appropriately;
- refactor _khmer.cc extension module to be more explicit about hashing.

No slowdown was observed on my hash function benchmarks (using #1436); if anything it
seemed about 5% faster _shrug_.

Note that Hashset still uses the global _hash function, as does KmerIterator; this doesn't
provide a one-stop shopping solution for replacing the global hash function :).
- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [x] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the `ChangeLog`?
  http://en.wikipedia.org/wiki/Changelog#Format
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)
- [x] Is the Copyright year up to date?

### CODE REVIEW

**Comment by ctb** [comment]
Ready for review @luizirber @betatim.

**Comment by betatim** [comment]
Is there a reason for having the various hash methods duplicated in `counting.hh` and `hashbits.hh` instead of putting them in `hashtable.hh`? Can we add a comment in `hashtable.hh` so we don't forget the reason?

I like the new name for `revhash`.

The left over direct calls to `_hash` could be redirected through `Hashtable::hash_dna(kmer)` as well if we make it a class method (`static` in c++ lingo). Would allow us to not have `convert_PyObject_to_HashIntoType` and `ht_convert_PyObject_to_HashIntoType` at the price of having to find a way to compute the k-size for each call. WDYT?

**Comment by betatim** [comment]
Nitpick: if we do keep `convert_PyObject_to_HashIntoType` and `ht_convert_PyObject_to_HashIntoType` can we put them next to each other in `_khmer.cc`. Would make it easier to compare them when you are wondering why there are two apparently equal functions ;)

File: lib/read_aligner.cc [hunk 1]
```diff
@@ -532,12 +532,13 @@ Alignment* ReadAligner::Align(const std::string& read)
     }
 
     HashIntoType fhash = 0, rhash = 0;
-    _hash(start.kmer, k, fhash, rhash);
+    fhash = m_ch->hash_dna_top_strand(start.kmer.c_str());
+    rhash = m_ch->hash_dna_bottom_strand(start.kmer.c_str());
```
Comment (betatim) [inline lib/read_aligner.cc]: Take a look at http://cpp.sh/5fce6 for something akin to "tuple unpacking" for return values. Might not be strictly necessary here but IMO makes for nice code especially compared to (ab)using arguments for return values 😃

**Comment by ctb** [comment]
Thanks for the comments, @betatim. I moved `convert_PyObject_to_HashIntoType` and `ht_convert_PyObject_to_HashIntoType` next to each other, and pulled `hash_dna` and `unhash_dna` back to `Hashtable`.

I'm punting on removing the last `_hash` calls from _khmer.cc and moving `hash_dna` to be class methods; I don't have a reason to go either way right now, and I think we will be revisiting the hash functions very soon in any case.

**Comment by ctb** [comment]
Actually, in further response to @betatim on the hash functions: in the not-so-distant future when we have multiple Hashtable-derived classes with different hash functions, how would we choose which hash function to use?  This is going to be a more general problem as we mix and match hashtables and hash functions... <sigh>

One solution might be to only get Hashsets from Hashtables, so they swipe the hash functions from there. An alternative might be to have two basic hash functions, one for irreversible and one for reversible, and just clearly distinguish their use in the code; then e.g. Hashset, Nodetable and Counttable would use irreversible, while Nodegraph and Countgraph would use reversible.

Both approaches sound like they'll get confusing... hmm.

Another idea is to have a global list of available hash functions and make sure they're specified explicitly when each object is created.  That could get ugly down the road but it works well with our standard modus operandi of primarily working through Python scripts.

**Comment by ctb** [comment]
Ready for review & merge @betatim

**Pull request merged**